Issue/7532 add utm params for card reader pages

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7532 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
In [Project Salas](https://github.com/woocommerce/woocommerce-ios/projects/51), we want to be able to understand the effectiveness of different JITM campaigns, initially those which link out to WooCommerce.com card reader purchase pages.

Adding UTM parameters to our links will allow us to attribute sales on WooCommerce.com, to the button in the apps the merchant tapped to get there.

This PR adds UTM parameters to the card reader purchase links.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Tap the CTA on the Upsell Card Reader banner (on the order list and payment methods screens) and the `Order card reader` row on the Payments menu.

Check that the following query parameters are added to the URL which is opened. 
Check that they are added to the `woocommercecom_product_view` tracks event.

- `utm_medium` – `woo_android`
- `utm_campaign` – `feature_announcement_card`, `payments_menu_item`: identifies whether this is a Feature Announcement Card, or a settings item.
- `utm_source` – `orders_list`, `payment_method`, or `payments_menu`
- `utm_term` – {STORE_ID}
- `utm_content` – `upsell_card_readers`: identifies the specific Feature Announcement Card which the tap came from. For other display locations, it's not required


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Initial commit that adds UTM provider class

This commit adds WooCommerceComUTMProvider.kt class which helps to transform the provided url with all the passed in utm query params.

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Small refactoring of WooCommerceComUTMProvider.kt class.

AnirudhBhat committed:
Integrate utm provider for ordering card reader link in the payments menu screen

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Integrate utm properties on purchase card reader click from the Order list screen

This commit integrates utm properties on purchase card reader click from the Order list screen on the upsell card reader banner.

AnirudhBhat committed:
Add tests to verify proper utm properties are populated in the hub screen

AnirudhBhat committed:
Integrate utm properties on purchase card reader click from the payment screen

This commit integrates utm properties on purchase card reader click from the Payment screen on the upsell card reader banner.

AnirudhBhat committed:
Add content utm property in Order list screen

AnirudhBhat committed:
Refactor WooCommerceComUTMProvider.kt class for easier readability

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Renamed test names for better readability and Added one more test case

AnirudhBhat committed:
Fix detekt errors

AnirudhBhat committed:
Added a comment to run WooCommerceComUTMProviderTest.kt manually when the behaviour of the WooCommerceComUTMProvider.kt class changes.

AnirudhBhat committed:
Added release notes

pachlava commented on RELEASE-NOTES.txt:
```diff
@@ -2,6 +2,7 @@
 10.8
 -----
 [*] [Internal] Signed-in card reader purchase from upsell banner both from the orders list and selection of a payment method (https://github.com/woocommerce/woocommerce-android/pull/7535)
+[*] [Internal] In-Person Payments: add UTM parameters to card reader purchase URLs to allow attribution (https://github.com/woocommerce/woocommerce-android/pull/7554)
```
Just leaving a note that this PR is targeting 10.9, so the release note should probably be listed under 10.9 🙂 

Also, the used format for GH links are square brackets (the one from above, 7535, also happens to use parentheses instead of square brackets).

AnirudhBhat replied:
Good catch! Fixed it [here](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/b4950053033b8cc19a5f773ff92432f5b7601148). Thank you!

pachlava reviewed (commented):
@AnirudhBhat 👋 

Apart from leaving [a note](https://github.com/woocommerce/woocommerce-android/pull/7554#discussion_r998130691) about release notes and seeing that newly added tests pass on FTL, the rest of my review was manual testing and checking that new parameters are indeed sent. And it looks like they are 👍 

`Order Card Reader` from IPP:

<img width="816" alt="ipp" src="https://user-images.githubusercontent.com/73365754/196444045-f99a9397-d44a-48fa-95b8-aad96e3201f4.png">

UCR banner, orders list:

<img width="1036" alt="list" src="https://user-images.githubusercontent.com/73365754/196444431-0923b909-b44b-4cf0-80e2-8627903f18ce.png">

UCR banner, payment method screen:

<img width="1066" alt="payment" src="https://user-images.githubusercontent.com/73365754/196444611-44091b4b-be12-4909-873e-68353e9f1358.png">

I've selected `Submit general feedback without explicit approval` only because of the release notes, which should be adjusted, IMO 🙂

kidinov commented on RELEASE-NOTES.txt:
```diff
@@ -2,6 +2,7 @@
 10.8
 -----
 [*] [Internal] Signed-in card reader purchase from upsell banner both from the orders list and selection of a payment method (https://github.com/woocommerce/woocommerce-android/pull/7535)
```
@AnirudhBhat That's my bad - I used `(` instead of `[` here. Could you please change it here as well? =)

AnirudhBhat replied:
No worries. Fixed [here](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/b4950053033b8cc19a5f773ff92432f5b7601148)

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooCommerceComUTMProvider.kt:
```diff
@@ -0,0 +1,90 @@
+     * utm_source=null -  is an invalid query
+     */
+    private fun isValidQuery(query: String): Boolean {
+        return !query.isNullOrEmpty() && (parameters[query]?.toString().isNullOrEmpty() || query !in parameters)
```
np: can be `query.isNotEmpty()`

AnirudhBhat replied:
Good suggestion. Fixed [here](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/fb558a2bb3d19025c9b7bf98f96113d762bd27b3)

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/SelectPaymentMethodFragment.kt:
```diff
@@ -212,7 +213,7 @@ class SelectPaymentMethodFragment : BaseFragment(R.layout.fragment_take_payment)
                     findNavController().navigate(
                         NavGraphMainDirections.actionGlobalWPComWebViewFragment(
-                            urlToLoad = event.url,
+                            urlToLoad = event.utmProvider.getUrlWithUtmParams(event.url.toUri()),
```
I am a bit confused with this approach. We are passing both the "mapper" and the "value" via event and then pass the value to the mapper

Won't be something like that more straightforward?

            val urlWithUtmParams = utmProvider.getUrlWithUtmParams(
                bannerDisplayEligibilityChecker.getPurchaseCardReaderUrl(source)
            )
            triggerEvent(
                OpenPurchaseCardReaderLink(
                    urlWithUtmParams,
                    R.string.card_reader_purchase_card_reader
                )
            )

AnirudhBhat replied:
The only reason passing in both the mapper and the value is because of the dependency on the `Uri` framework class. 

```
val urlWithUtmParams = utmProvider.getUrlWithUtmParams(
            bannerDisplayEligibilityChecker.getPurchaseCardReaderUrl(source)
        )
        triggerEvent(
            OpenPurchaseCardReaderLink(
                urlWithUtmParams,
                R.string.card_reader_purchase_card_reader
            )
        )
```

This ☝️ approach would cause the unit test to fail since it internally depends on the `Uri` framework class.

Now with [this](https://github.com/woocommerce/woocommerce-android/pull/7554#discussion_r998992376) approach, this is no longer a concern. We are good here.

kidinov replied:
Yep, I forgot to mention that there was not a real big need to have `Uri` as a parameter

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/WooCommerceComUTMProvider.kt:
```diff
@@ -0,0 +1,90 @@
+ * WooCommerceComUTMProviderTest will not be run as part of PR checks, it's important to run it manually when
+ * the behaviour of this class changes.
+ */
+@Reusable
```
It seems to be injectable, but we never inject this. Maybe to have 3 named injectable `WooCommerceComUTMProvider`:

* OrderListUpselCardReaderUtmProvider
* SelectPaymentMethodUpselCardReaderUtmProvider
* etc.. 


Probably will be easier to test and reuse

AnirudhBhat replied:
I like this suggestion! Implemented it [here](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/78a0349950c4196c2e3bd0f5994b79235cdc7038) and small refactor [here](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/fb558a2bb3d19025c9b7bf98f96113d762bd27b3). Thank you!

kidinov reviewed (commented):
Thank you for working on that! Looks good. I left a couple of comments to consider

AnirudhBhat committed:
Refactor UtmProvider.kt interface

Remove WooCommerceComUTMProvider.kt class and move all the core logic to the UtmProvider.kt interface. All the different utm providers are now injected as dependency to whoever needs.

AnirudhBhat committed:
Make UtmProvider.kt an abstract class instead of interface

AnirudhBhat committed:
Fix markdown issue and add release notes under appropriate release version

AnirudhBhat committed:
Merge branch 'trunk' into issue/7532-add-utm-params-for-card-reader-pages

kidinov commented on RELEASE-NOTES.txt:
```diff
@@ -1,8 +1,11 @@
 -----
-[*] [Internal] Signed-in card reader purchase from upsell banner both from the orders list and selection of a payment method (https://github.com/woocommerce/woocommerce-android/pull/7535)
-[*] [Internal] In-Person Payments: add UTM parameters to card reader purchase URLs to allow attribution (https://github.com/woocommerce/woocommerce-android/pull/7554)
+[*] [Internal] Signed-in card reader purchase from upsell banner both from the orders list and selection of a payment method [https://github.com/woocommerce/woocommerce-android/pull/7535]
```
🙏

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/hub/CardReaderHubViewModel.kt:
```diff
@@ -248,16 +249,10 @@ class CardReaderHubViewModel @Inject constructor(
-            content = null,
-            siteId = selectedSite.getIfExists()?.siteId
-        )
+        println(paymentMenuUtmProvider.getUrlWithUtmParams(cardReaderPurchaseUrl))
```
Probably leftover from debuging?

AnirudhBhat replied:
Oops! Fixed [here](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/c1f1eb7ac32de13feb7e4c209c86e56958b045e0). Thank you!

kidinov reviewed (approved):
LGTM!

> https://woocommerce.com/products/hardware/US?utm_campaign=feature_announcement_card&utm_source=orders_list&utm_content=upsell_card_readers&utm_term=192152755&utm_medium=woo_android
https://woocommerce.com/products/hardware/US?utm_campaign=feature_announcement_card&utm_source=payment_method&utm_content=upsell_card_readers&utm_term=192152755&utm_medium=woo_android
https://woocommerce.com/products/hardware/US?utm_campaign=payments_menu_item&utm_source=payments_menu&utm_term=192152755&utm_medium=woo_android


You probably just want to remove "println" before merging

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/UtmProvider.kt:
```diff
@@ -0,0 +1,80 @@
+ * WooCommerceComUTMProviderTest will not be run as part of PR checks, it's important to run it manually when
+ * the behaviour of this class changes.
+ */
+abstract class UtmProvider {
```
Just out of curiosity, thinking out loud - what's the benefit of defining `UtmProvider` as an `abstract` class? 

Would it work to make it a `final` class and pass those properties in a constructor, e.g.:
```kotlin
class UtmProvider(
    val campaign: String, 
    val source: String, 
    val content: String?, 
    val siteId: Long?) {
//... 
}
```

and in `WooCommerceComUTMProviderModule` we could construct instances like this:
```kotlin
@Provides
@Singleton
@Named("payment-menu")
    fun providePaymentMenuUtm(
        selectedSite: SelectedSite
    ) = UtmProvider(
        CardReaderHubViewModel.UTM_CAMPAIGN,
        CardReaderHubViewModel.UTM_SOURCE,
        null,
        selectedSite.getIfExists()?.siteId
    )
```

But, maybe I'm missing something, anyway I prefer simplifying things 🙂

AnirudhBhat replied:
Agree! [Done](https://github.com/woocommerce/woocommerce-android/pull/7554/commits/04990b070ed91dbdcaa40c2e3d96a32fac5bf106)

AnirudhBhat committed:
Make UtmProvider.kt final class instead of abstract class

AnirudhBhat committed:
Remove log used for debugging
